<!DOCTYPE html>
<!-- Generated by pkgdown: do not edit by hand --><html lang="en">
<head>
<meta http-equiv="Content-Type" content="text/html; charset=UTF-8">
<meta charset="utf-8">
<meta http-equiv="X-UA-Compatible" content="IE=edge">
<meta name="viewport" content="width=device-width, initial-scale=1, shrink-to-fit=no">
<meta name="description" content="nfportalutils">
<title>Revalidation workflows • nfportalutils</title>
<script src="../deps/jquery-3.6.0/jquery-3.6.0.min.js"></script><meta name="viewport" content="width=device-width, initial-scale=1, shrink-to-fit=no">
<link href="../deps/bootstrap-5.3.1/bootstrap.min.css" rel="stylesheet">
<script src="../deps/bootstrap-5.3.1/bootstrap.bundle.min.js"></script><!-- Font Awesome icons --><link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/all.min.css" integrity="sha256-mmgLkCYLUQbXn0B1SRqzHar6dCnv9oZFPEC1g1cwlkk=" crossorigin="anonymous">
<link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/v4-shims.min.css" integrity="sha256-wZjR52fzng1pJHwx4aV2AO3yyTOXrcDW7jBpJtTwVxw=" crossorigin="anonymous">
<!-- bootstrap-toc --><script src="https://cdn.jsdelivr.net/gh/afeld/bootstrap-toc@v1.0.1/dist/bootstrap-toc.min.js" integrity="sha256-4veVQbu7//Lk5TSmc7YV48MxtMy98e26cf5MrgZYnwo=" crossorigin="anonymous"></script><!-- headroom.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/headroom.min.js" integrity="sha256-AsUX4SJE1+yuDu5+mAVzJbuYNPHj/WroHuZ8Ir/CkE0=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/jQuery.headroom.min.js" integrity="sha256-ZX/yNShbjqsohH1k95liqY9Gd8uOiE1S4vZc+9KQ1K4=" crossorigin="anonymous"></script><!-- clipboard.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/clipboard.js/2.0.6/clipboard.min.js" integrity="sha256-inc5kl9MA1hkeYUt+EC3BhlIgyp/2jDIyBLS6k3UxPI=" crossorigin="anonymous"></script><!-- search --><script src="https://cdnjs.cloudflare.com/ajax/libs/fuse.js/6.4.6/fuse.js" integrity="sha512-zv6Ywkjyktsohkbp9bb45V6tEMoWhzFzXis+LrMehmJZZSys19Yxf1dopHx7WzIKxr5tK2dVcYmaCk2uqdjF4A==" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/autocomplete.js/0.38.0/autocomplete.jquery.min.js" integrity="sha512-GU9ayf+66Xx2TmpxqJpliWbT5PiGYxpaG8rfnBEk1LL8l1KGkRShhngwdXK1UgqhAzWpZHSiYPc09/NwDQIGyg==" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/mark.js/8.11.1/mark.min.js" integrity="sha512-5CYOlHXGh6QpOFA/TeTylKLWfB3ftPsde7AnmhuitiTX4K5SqCLBeKro6sPS8ilsz1Q4NRx3v8Ko2IBiszzdww==" crossorigin="anonymous"></script><!-- pkgdown --><script src="../pkgdown.js"></script><meta property="og:title" content="Revalidation workflows">
<meta property="og:description" content="nfportalutils">
<!-- mathjax --><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/MathJax.js" integrity="sha256-nvJJv9wWKEm88qvoQl9ekL2J+k/RWIsaSScxxlsrv8k=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/config/TeX-AMS-MML_HTMLorMML.js" integrity="sha256-84DKXVJXs0/F8OTMzX4UR909+jtl4G7SPypPavF+GfA=" crossorigin="anonymous"></script><!--[if lt IE 9]>
<script src="https://oss.maxcdn.com/html5shiv/3.7.3/html5shiv.min.js"></script>
<script src="https://oss.maxcdn.com/respond/1.4.2/respond.min.js"></script>
<![endif]-->
</head>
<body>
    <a href="#main" class="visually-hidden-focusable">Skip to contents</a>
    

    <nav class="navbar fixed-top navbar-light navbar-expand-lg bg-light"><div class="container">
    
    <a class="navbar-brand me-2" href="../index.html">nfportalutils</a>

    <small class="nav-text text-muted me-auto" data-bs-toggle="tooltip" data-bs-placement="bottom" title="">0.9500</small>

    
    <button class="navbar-toggler" type="button" data-bs-toggle="collapse" data-bs-target="#navbar" aria-controls="navbar" aria-expanded="false" aria-label="Toggle navigation">
      <span class="navbar-toggler-icon"></span>
    </button>

    <div id="navbar" class="collapse navbar-collapse ms-3">
      <ul class="navbar-nav me-auto">
<li class="nav-item">
  <a class="nav-link" href="../reference/index.html">Reference</a>
</li>
<li class="active nav-item dropdown">
  <a href="#" class="nav-link dropdown-toggle" data-bs-toggle="dropdown" role="button" aria-expanded="false" aria-haspopup="true" id="dropdown-articles">Articles</a>
  <div class="dropdown-menu" aria-labelledby="dropdown-articles">
    <a class="dropdown-item" href="../articles/annotate-nf-processed-data.html">Annotating nextflow processed data</a>
    <a class="dropdown-item" href="../articles/bringing-portal-data-to-other-platforms-cbioportal.html">Bringing Portal Data to Other Platforms: cBioPortal</a>
    <a class="dropdown-item" href="../articles/revalidation-workflows.html">Revalidation workflows</a>
    <a class="dropdown-item" href="../articles/survey-public-files.html">Surveying public files in the portal</a>
  </div>
</li>
      </ul>
<form class="form-inline my-2 my-lg-0" role="search">
        <input type="search" class="form-control me-sm-2" aria-label="Toggle navigation" name="search-input" data-search-index="../search.json" id="search-input" placeholder="Search for" autocomplete="off">
</form>

      <ul class="navbar-nav">
<li class="nav-item">
  <a class="external-link nav-link" href="https://github.com/nf-osi/nfportalutils/" aria-label="github">
    <span class="fab fa fab fa-github fa-lg"></span>
     
  </a>
</li>
      </ul>
</div>

    
  </div>
</nav><div class="container template-article">




<div class="row">
  <main id="main" class="col-md-9"><div class="page-header">
      <img src="" class="logo" alt=""><h1>Revalidation workflows</h1>
            
      
      <small class="dont-index">Source: <a href="https://github.com/nf-osi/nfportalutils/blob/HEAD/vignettes/revalidation-workflows.Rmd" class="external-link"><code>vignettes/revalidation-workflows.Rmd</code></a></small>
      <div class="d-none name"><code>revalidation-workflows.Rmd</code></div>
    </div>

    
    
<p>This vignette demonstrates variants of metadata revalidation
workflows.</p>
<p>First set up as usual.</p>
<div class="sourceCode" id="cb1"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span></span>
<span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va"><a href="https://github.com/nf-osi/nfportalutils" class="external-link">nfportalutils</a></span><span class="op">)</span></span>
<span><span class="fu"><a href="../reference/syn_login.html">syn_login</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/Sys.getenv.html" class="external-link">Sys.getenv</a></span><span class="op">(</span><span class="st">"SYNAPSE_AUTH_TOKEN"</span><span class="op">)</span><span class="op">)</span></span></code></pre></div>
<div class="section level2">
<h2 id="basics-with-schematic-api-service">Basics with Schematic API service<a class="anchor" aria-label="anchor" href="#basics-with-schematic-api-service"></a>
</h2>
<p>Schematic API only works with dataset folders currently. Find a
dataset folder.</p>
<ol style="list-style-type: decimal">
<li>To validate metadata, a manifest must be reconstituted. Type
<code><a href="../reference/manifest_generate.html">?manifest_generate</a></code> to read the docs.</li>
</ol>
<p>As seen in the params below, you do need to know the
<code>data_type</code> to validate against. The <code>data_type</code>
is the same as the “Component” in the schematic data model (and so the
exact term depends on the data model). If feeling lucky, try
<code>infer_data_type</code>.</p>
<div class="sourceCode" id="cb2"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span></span>
<span><span class="va">my_dataset</span> <span class="op">&lt;-</span> <span class="st">"syn25386362"</span></span>
<span></span>
<span><span class="va">inferred</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/infer_data_type.html">infer_data_type</a></span><span class="op">(</span><span class="va">my_dataset</span><span class="op">)</span></span>
<span><span class="va">inferred</span></span>
<span><span class="co">#&gt; $result</span></span>
<span><span class="co">#&gt; [1] "GenomicsAssayTemplate"</span></span>
<span><span class="co">#&gt; </span></span>
<span><span class="co">#&gt; $notes</span></span>
<span><span class="co">#&gt; [1] ""</span></span>
<span></span>
<span><span class="va">data_type</span> <span class="op">&lt;-</span> <span class="va">inferred</span><span class="op">$</span><span class="va">result</span></span>
<span></span>
<span><span class="fu"><a href="../reference/manifest_generate.html">manifest_generate</a></span><span class="op">(</span><span class="va">data_type</span>, </span>
<span>                  dataset_id <span class="op">=</span> <span class="va">my_dataset</span>,</span>
<span>                  schema_url <span class="op">=</span> <span class="st">"https://raw.githubusercontent.com/nf-osi/nf-metadata-dictionary/main/NF.jsonld"</span>,</span>
<span>                  output_format <span class="op">=</span> <span class="st">"google_sheet"</span><span class="op">)</span> <span class="co"># otherwise excel </span></span>
<span><span class="co">#&gt; Manifest generated as Googlesheet(s)</span></span>
<span><span class="co">#&gt; [[1]]</span></span>
<span><span class="co">#&gt; [1] "https://docs.google.com/spreadsheets/d/11ymlnESzn7XhHS3vHzlRFsDPhafJJehlhuCL9HgICt8"</span></span></code></pre></div>
<ol start="2" style="list-style-type: decimal">
<li>Go to google_sheet and download as <code>.csv</code>. If Excel was
chosen, open in some spreadsheet editor and resave file as
<code>.csv</code>. Then validate.</li>
</ol>
<div class="sourceCode" id="cb3"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="../reference/manifest_validate.html">manifest_validate</a></span><span class="op">(</span>data_type <span class="op">=</span> <span class="va">data_type</span>, </span>
<span>                  file_name <span class="op">=</span> <span class="st">"GenomicsAssayTemplate - Sheet1.csv"</span><span class="op">)</span></span>
<span><span class="co">#&gt; $errors</span></span>
<span><span class="co">#&gt; $errors[[1]]</span></span>
<span><span class="co">#&gt; $errors[[1]][[1]]</span></span>
<span><span class="co">#&gt; [1] "2"</span></span>
<span><span class="co">#&gt; </span></span>
<span><span class="co">#&gt; $errors[[1]][[2]]</span></span>
<span><span class="co">#&gt; [1] "assay"</span></span>
<span><span class="co">#&gt; </span></span>
<span><span class="co">#&gt; $errors[[1]][[3]]</span></span>
<span><span class="co">#&gt; [1] "'' is not one of ['NIH Toolbox', 'STR profile', 'traction force microscopy', 'massively parallel reporter assay', 'gait measurement', 'conventional MRI', 'functional MRI', 'immunoassay', 'contextual conditioning behavior assay', 'genotyping', 'DNA optical mapping', 'NOMe-seq', 'Social Responsiveness Scale', 'targeted exome sequencing', '2D AlamarBlue fluorescence', 'TMT quantitation', 'liquid chromatography-electrochemical detection', 'whole genome sequencing', 'Riccardi and Ablon scales', 'cell"</span></span>
<span><span class="co">#&gt; </span></span>
<span><span class="co">#&gt; $errors[[1]][[4]]</span></span>
<span><span class="co">#&gt; [1] ""</span></span>
<span><span class="co">#&gt; </span></span>
<span><span class="co">#&gt; </span></span>
<span><span class="co">#&gt; $errors[[2]]</span></span>
<span><span class="co">#&gt; $errors[[2]][[1]]</span></span>
<span><span class="co">#&gt; [1] "2"</span></span>
<span><span class="co">#&gt; </span></span>
<span><span class="co">#&gt; $errors[[2]][[2]]</span></span>
<span><span class="co">#&gt; [1] "specimenID"</span></span>
<span><span class="co">#&gt; </span></span>
<span><span class="co">#&gt; $errors[[2]][[3]]</span></span>
<span><span class="co">#&gt; [1] "'' is too short"</span></span>
<span><span class="co">#&gt; </span></span>
<span><span class="co">#&gt; $errors[[2]][[4]]</span></span>
<span><span class="co">#&gt; [1] ""</span></span>
<span><span class="co">#&gt; </span></span>
<span><span class="co">#&gt; </span></span>
<span><span class="co">#&gt; $errors[[3]]</span></span>
<span><span class="co">#&gt; $errors[[3]][[1]]</span></span>
<span><span class="co">#&gt; [1] "2"</span></span>
<span><span class="co">#&gt; </span></span>
<span><span class="co">#&gt; $errors[[3]][[2]]</span></span>
<span><span class="co">#&gt; [1] "libraryStrand"</span></span>
<span><span class="co">#&gt; </span></span>
<span><span class="co">#&gt; $errors[[3]][[3]]</span></span>
<span><span class="co">#&gt; [1] "'' is not one of ['FirstStranded', 'Unstranded', 'Not Applicable', 'SecondStranded']"</span></span>
<span><span class="co">#&gt; </span></span>
<span><span class="co">#&gt; $errors[[3]][[4]]</span></span>
<span><span class="co">#&gt; [1] ""</span></span>
<span><span class="co">#&gt; </span></span>
<span><span class="co">#&gt; </span></span>
<span><span class="co">#&gt; $errors[[4]]</span></span>
<span><span class="co">#&gt; $errors[[4]][[1]]</span></span>
<span><span class="co">#&gt; [1] "2"</span></span>
<span><span class="co">#&gt; </span></span>
<span><span class="co">#&gt; $errors[[4]][[2]]</span></span>
<span><span class="co">#&gt; [1] "tumorType"</span></span>
<span><span class="co">#&gt; </span></span>
<span><span class="co">#&gt; $errors[[4]][[3]]</span></span>
<span><span class="co">#&gt; [1] "'' is not one of ['Anaplastic Ganglioglioma', 'Anaplastic Astrocytoma', 'Nodular Neurofibroma', 'Meningioma', 'Fibrosarcoma', 'Localized Neurofibroma', 'Glioblastoma', 'Malignant Peripheral Nerve Sheath Tumor', 'Anaplastic Pleomorphic Xanthoastrocytoma', 'Atypical Neurofibroma', 'tumor', 'Colorectal Adenocarcinoma', 'Recurrent MPNST', 'Pilocytic Astrocytoma', 'Ganglioglioma', 'Optic Pathway Glioma', 'Neurofibroma', 'Necrotic Neoplasm', 'Glioma', 'Teratoma', 'Cutaneous Neurofibroma', 'Fibromatosi"</span></span>
<span><span class="co">#&gt; </span></span>
<span><span class="co">#&gt; $errors[[4]][[4]]</span></span>
<span><span class="co">#&gt; [1] ""</span></span>
<span><span class="co">#&gt; </span></span>
<span><span class="co">#&gt; </span></span>
<span><span class="co">#&gt; $errors[[5]]</span></span>
<span><span class="co">#&gt; $errors[[5]][[1]]</span></span>
<span><span class="co">#&gt; [1] "2"</span></span>
<span><span class="co">#&gt; </span></span>
<span><span class="co">#&gt; $errors[[5]][[2]]</span></span>
<span><span class="co">#&gt; [1] "libraryPreparationMethod"</span></span>
<span><span class="co">#&gt; </span></span>
<span><span class="co">#&gt; $errors[[5]][[3]]</span></span>
<span><span class="co">#&gt; [1] "'' is not one of ['CEL-seq', 'NEBNext mRNA Library Prep Reagent Set for Illumina', '10x', 'GTAC@WUSTL in-house prep', 'KAPA mRNA HyperPrep Kit', 'TruSeq', 'unknown', 'KAPA HyperPrep Kit PCR-free', 'Illumina TruSeq DNA Nano', 'TruSeq standard total RNA library kit', 'QuantSeq FWD V2 with UDI', 'Drop-Seq', 'KAPA RNA HyperPrep Kit with RiboErase (HMR)', 'Smart-seq4', 'IDT xGen Exome Research Panel', 'Smart-seq2', 'Omni-ATAC']"</span></span>
<span><span class="co">#&gt; </span></span>
<span><span class="co">#&gt; $errors[[5]][[4]]</span></span>
<span><span class="co">#&gt; [1] ""</span></span>
<span><span class="co">#&gt; </span></span>
<span><span class="co">#&gt; </span></span>
<span><span class="co">#&gt; $errors[[6]]</span></span>
<span><span class="co">#&gt; $errors[[6]][[1]]</span></span>
<span><span class="co">#&gt; [1] "2"</span></span>
<span><span class="co">#&gt; </span></span>
<span><span class="co">#&gt; $errors[[6]][[2]]</span></span>
<span><span class="co">#&gt; [1] "individualID"</span></span>
<span><span class="co">#&gt; </span></span>
<span><span class="co">#&gt; $errors[[6]][[3]]</span></span>
<span><span class="co">#&gt; [1] "'' is too short"</span></span>
<span><span class="co">#&gt; </span></span>
<span><span class="co">#&gt; $errors[[6]][[4]]</span></span>
<span><span class="co">#&gt; [1] ""</span></span>
<span><span class="co">#&gt; </span></span>
<span><span class="co">#&gt; </span></span>
<span><span class="co">#&gt; $errors[[7]]</span></span>
<span><span class="co">#&gt; $errors[[7]][[1]]</span></span>
<span><span class="co">#&gt; [1] "2"</span></span>
<span><span class="co">#&gt; </span></span>
<span><span class="co">#&gt; $errors[[7]][[2]]</span></span>
<span><span class="co">#&gt; [1] "platform"</span></span>
<span><span class="co">#&gt; </span></span>
<span><span class="co">#&gt; $errors[[7]][[3]]</span></span>
<span><span class="co">#&gt; [1] "'' is not one of ['Illumina Genome Analyzer IIx', 'Illumina HiSeq X', 'Perlegen 300Karray', 'Vevo 3100 Imaging System', 'Illumina MouseWG-6 v2.0 expression beadchip', 'Illumina Infinium MethylationEPIC BeadChip v2.0 (935k)', 'Vectra H1 3D Imaging System', 'Nanostring Counter', 'Illumina Infinium MethylationEPIC BeadChip v1.0 (850k)', 'Illumina HumanOmniExpress-24 v1.0 BeadChip', 'Illumina HumanOmni1-Quadv1.0', 'LifeViz Micro System', 'LI-COR Odyssey CLx', 'Illumina HumanMethylation450', 'Illumin"</span></span>
<span><span class="co">#&gt; </span></span>
<span><span class="co">#&gt; $errors[[7]][[4]]</span></span>
<span><span class="co">#&gt; [1] ""</span></span>
<span><span class="co">#&gt; </span></span>
<span><span class="co">#&gt; </span></span>
<span><span class="co">#&gt; $errors[[8]]</span></span>
<span><span class="co">#&gt; $errors[[8]][[1]]</span></span>
<span><span class="co">#&gt; [1] "2"</span></span>
<span><span class="co">#&gt; </span></span>
<span><span class="co">#&gt; $errors[[8]][[2]]</span></span>
<span><span class="co">#&gt; [1] "specimenPreparationMethod"</span></span>
<span><span class="co">#&gt; </span></span>
<span><span class="co">#&gt; $errors[[8]][[3]]</span></span>
<span><span class="co">#&gt; [1] "'' is not one of ['FFPE', 'OCT', 'RNAlater', 'Viably frozen', 'Fresh collected', 'Cryopreserved', 'formalin-fixed', 'Flash frozen', 'ethanol']"</span></span>
<span><span class="co">#&gt; </span></span>
<span><span class="co">#&gt; $errors[[8]][[4]]</span></span>
<span><span class="co">#&gt; [1] ""</span></span>
<span><span class="co">#&gt; </span></span>
<span><span class="co">#&gt; </span></span>
<span><span class="co">#&gt; $errors[[9]]</span></span>
<span><span class="co">#&gt; $errors[[9]][[1]]</span></span>
<span><span class="co">#&gt; [1] "2"</span></span>
<span><span class="co">#&gt; </span></span>
<span><span class="co">#&gt; $errors[[9]][[2]]</span></span>
<span><span class="co">#&gt; [1] "species"</span></span>
<span><span class="co">#&gt; </span></span>
<span><span class="co">#&gt; $errors[[9]][[3]]</span></span>
<span><span class="co">#&gt; [1] "'' is not one of ['Rattus norvegicus', 'Gallus gallus', 'Danio rerio', 'Sus scrofa', 'Drosophila melanogaster', 'Oryctolagus cuniculus', 'Pan troglodytes', 'Rhesus macaque', 'Mus musculus (humanized)', 'Homo sapiens', 'Mus musculus']"</span></span>
<span><span class="co">#&gt; </span></span>
<span><span class="co">#&gt; $errors[[9]][[4]]</span></span>
<span><span class="co">#&gt; [1] ""</span></span>
<span><span class="co">#&gt; </span></span>
<span><span class="co">#&gt; </span></span>
<span><span class="co">#&gt; $errors[[10]]</span></span>
<span><span class="co">#&gt; $errors[[10]][[1]]</span></span>
<span><span class="co">#&gt; [1] "3"</span></span>
<span><span class="co">#&gt; </span></span>
<span><span class="co">#&gt; $errors[[10]][[2]]</span></span>
<span><span class="co">#&gt; [1] "assay"</span></span>
<span><span class="co">#&gt; </span></span>
<span><span class="co">#&gt; $errors[[10]][[3]]</span></span>
<span><span class="co">#&gt; [1] "'' is not one of ['NIH Toolbox', 'STR profile', 'traction force microscopy', 'massively parallel reporter assay', 'gait measurement', 'conventional MRI', 'functional MRI', 'immunoassay', 'contextual conditioning behavior assay', 'genotyping', 'DNA optical mapping', 'NOMe-seq', 'Social Responsiveness Scale', 'targeted exome sequencing', '2D AlamarBlue fluorescence', 'TMT quantitation', 'liquid chromatography-electrochemical detection', 'whole genome sequencing', 'Riccardi and Ablon scales', 'cell"</span></span>
<span><span class="co">#&gt; </span></span>
<span><span class="co">#&gt; $errors[[10]][[4]]</span></span>
<span><span class="co">#&gt; [1] ""</span></span>
<span><span class="co">#&gt; </span></span>
<span><span class="co">#&gt; </span></span>
<span><span class="co">#&gt; $errors[[11]]</span></span>
<span><span class="co">#&gt; $errors[[11]][[1]]</span></span>
<span><span class="co">#&gt; [1] "3"</span></span>
<span><span class="co">#&gt; </span></span>
<span><span class="co">#&gt; $errors[[11]][[2]]</span></span>
<span><span class="co">#&gt; [1] "specimenID"</span></span>
<span><span class="co">#&gt; </span></span>
<span><span class="co">#&gt; $errors[[11]][[3]]</span></span>
<span><span class="co">#&gt; [1] "'' is too short"</span></span>
<span><span class="co">#&gt; </span></span>
<span><span class="co">#&gt; $errors[[11]][[4]]</span></span>
<span><span class="co">#&gt; [1] ""</span></span>
<span><span class="co">#&gt; </span></span>
<span><span class="co">#&gt; </span></span>
<span><span class="co">#&gt; $errors[[12]]</span></span>
<span><span class="co">#&gt; $errors[[12]][[1]]</span></span>
<span><span class="co">#&gt; [1] "3"</span></span>
<span><span class="co">#&gt; </span></span>
<span><span class="co">#&gt; $errors[[12]][[2]]</span></span>
<span><span class="co">#&gt; [1] "libraryStrand"</span></span>
<span><span class="co">#&gt; </span></span>
<span><span class="co">#&gt; $errors[[12]][[3]]</span></span>
<span><span class="co">#&gt; [1] "'' is not one of ['FirstStranded', 'Unstranded', 'Not Applicable', 'SecondStranded']"</span></span>
<span><span class="co">#&gt; </span></span>
<span><span class="co">#&gt; $errors[[12]][[4]]</span></span>
<span><span class="co">#&gt; [1] ""</span></span>
<span><span class="co">#&gt; </span></span>
<span><span class="co">#&gt; </span></span>
<span><span class="co">#&gt; $errors[[13]]</span></span>
<span><span class="co">#&gt; $errors[[13]][[1]]</span></span>
<span><span class="co">#&gt; [1] "3"</span></span>
<span><span class="co">#&gt; </span></span>
<span><span class="co">#&gt; $errors[[13]][[2]]</span></span>
<span><span class="co">#&gt; [1] "tumorType"</span></span>
<span><span class="co">#&gt; </span></span>
<span><span class="co">#&gt; $errors[[13]][[3]]</span></span>
<span><span class="co">#&gt; [1] "'' is not one of ['Anaplastic Ganglioglioma', 'Anaplastic Astrocytoma', 'Nodular Neurofibroma', 'Meningioma', 'Fibrosarcoma', 'Localized Neurofibroma', 'Glioblastoma', 'Malignant Peripheral Nerve Sheath Tumor', 'Anaplastic Pleomorphic Xanthoastrocytoma', 'Atypical Neurofibroma', 'tumor', 'Colorectal Adenocarcinoma', 'Recurrent MPNST', 'Pilocytic Astrocytoma', 'Ganglioglioma', 'Optic Pathway Glioma', 'Neurofibroma', 'Necrotic Neoplasm', 'Glioma', 'Teratoma', 'Cutaneous Neurofibroma', 'Fibromatosi"</span></span>
<span><span class="co">#&gt; </span></span>
<span><span class="co">#&gt; $errors[[13]][[4]]</span></span>
<span><span class="co">#&gt; [1] ""</span></span>
<span><span class="co">#&gt; </span></span>
<span><span class="co">#&gt; </span></span>
<span><span class="co">#&gt; $errors[[14]]</span></span>
<span><span class="co">#&gt; $errors[[14]][[1]]</span></span>
<span><span class="co">#&gt; [1] "3"</span></span>
<span><span class="co">#&gt; </span></span>
<span><span class="co">#&gt; $errors[[14]][[2]]</span></span>
<span><span class="co">#&gt; [1] "libraryPreparationMethod"</span></span>
<span><span class="co">#&gt; </span></span>
<span><span class="co">#&gt; $errors[[14]][[3]]</span></span>
<span><span class="co">#&gt; [1] "'' is not one of ['CEL-seq', 'NEBNext mRNA Library Prep Reagent Set for Illumina', '10x', 'GTAC@WUSTL in-house prep', 'KAPA mRNA HyperPrep Kit', 'TruSeq', 'unknown', 'KAPA HyperPrep Kit PCR-free', 'Illumina TruSeq DNA Nano', 'TruSeq standard total RNA library kit', 'QuantSeq FWD V2 with UDI', 'Drop-Seq', 'KAPA RNA HyperPrep Kit with RiboErase (HMR)', 'Smart-seq4', 'IDT xGen Exome Research Panel', 'Smart-seq2', 'Omni-ATAC']"</span></span>
<span><span class="co">#&gt; </span></span>
<span><span class="co">#&gt; $errors[[14]][[4]]</span></span>
<span><span class="co">#&gt; [1] ""</span></span>
<span><span class="co">#&gt; </span></span>
<span><span class="co">#&gt; </span></span>
<span><span class="co">#&gt; $errors[[15]]</span></span>
<span><span class="co">#&gt; $errors[[15]][[1]]</span></span>
<span><span class="co">#&gt; [1] "3"</span></span>
<span><span class="co">#&gt; </span></span>
<span><span class="co">#&gt; $errors[[15]][[2]]</span></span>
<span><span class="co">#&gt; [1] "individualID"</span></span>
<span><span class="co">#&gt; </span></span>
<span><span class="co">#&gt; $errors[[15]][[3]]</span></span>
<span><span class="co">#&gt; [1] "'' is too short"</span></span>
<span><span class="co">#&gt; </span></span>
<span><span class="co">#&gt; $errors[[15]][[4]]</span></span>
<span><span class="co">#&gt; [1] ""</span></span>
<span><span class="co">#&gt; </span></span>
<span><span class="co">#&gt; </span></span>
<span><span class="co">#&gt; $errors[[16]]</span></span>
<span><span class="co">#&gt; $errors[[16]][[1]]</span></span>
<span><span class="co">#&gt; [1] "3"</span></span>
<span><span class="co">#&gt; </span></span>
<span><span class="co">#&gt; $errors[[16]][[2]]</span></span>
<span><span class="co">#&gt; [1] "platform"</span></span>
<span><span class="co">#&gt; </span></span>
<span><span class="co">#&gt; $errors[[16]][[3]]</span></span>
<span><span class="co">#&gt; [1] "'' is not one of ['Illumina Genome Analyzer IIx', 'Illumina HiSeq X', 'Perlegen 300Karray', 'Vevo 3100 Imaging System', 'Illumina MouseWG-6 v2.0 expression beadchip', 'Illumina Infinium MethylationEPIC BeadChip v2.0 (935k)', 'Vectra H1 3D Imaging System', 'Nanostring Counter', 'Illumina Infinium MethylationEPIC BeadChip v1.0 (850k)', 'Illumina HumanOmniExpress-24 v1.0 BeadChip', 'Illumina HumanOmni1-Quadv1.0', 'LifeViz Micro System', 'LI-COR Odyssey CLx', 'Illumina HumanMethylation450', 'Illumin"</span></span>
<span><span class="co">#&gt; </span></span>
<span><span class="co">#&gt; $errors[[16]][[4]]</span></span>
<span><span class="co">#&gt; [1] ""</span></span>
<span><span class="co">#&gt; </span></span>
<span><span class="co">#&gt; </span></span>
<span><span class="co">#&gt; $errors[[17]]</span></span>
<span><span class="co">#&gt; $errors[[17]][[1]]</span></span>
<span><span class="co">#&gt; [1] "3"</span></span>
<span><span class="co">#&gt; </span></span>
<span><span class="co">#&gt; $errors[[17]][[2]]</span></span>
<span><span class="co">#&gt; [1] "specimenPreparationMethod"</span></span>
<span><span class="co">#&gt; </span></span>
<span><span class="co">#&gt; $errors[[17]][[3]]</span></span>
<span><span class="co">#&gt; [1] "'' is not one of ['FFPE', 'OCT', 'RNAlater', 'Viably frozen', 'Fresh collected', 'Cryopreserved', 'formalin-fixed', 'Flash frozen', 'ethanol']"</span></span>
<span><span class="co">#&gt; </span></span>
<span><span class="co">#&gt; $errors[[17]][[4]]</span></span>
<span><span class="co">#&gt; [1] ""</span></span>
<span><span class="co">#&gt; </span></span>
<span><span class="co">#&gt; </span></span>
<span><span class="co">#&gt; $errors[[18]]</span></span>
<span><span class="co">#&gt; $errors[[18]][[1]]</span></span>
<span><span class="co">#&gt; [1] "3"</span></span>
<span><span class="co">#&gt; </span></span>
<span><span class="co">#&gt; $errors[[18]][[2]]</span></span>
<span><span class="co">#&gt; [1] "species"</span></span>
<span><span class="co">#&gt; </span></span>
<span><span class="co">#&gt; $errors[[18]][[3]]</span></span>
<span><span class="co">#&gt; [1] "'' is not one of ['Rattus norvegicus', 'Gallus gallus', 'Danio rerio', 'Sus scrofa', 'Drosophila melanogaster', 'Oryctolagus cuniculus', 'Pan troglodytes', 'Rhesus macaque', 'Mus musculus (humanized)', 'Homo sapiens', 'Mus musculus']"</span></span>
<span><span class="co">#&gt; </span></span>
<span><span class="co">#&gt; $errors[[18]][[4]]</span></span>
<span><span class="co">#&gt; [1] ""</span></span>
<span><span class="co">#&gt; </span></span>
<span><span class="co">#&gt; </span></span>
<span><span class="co">#&gt; </span></span>
<span><span class="co">#&gt; $warnings</span></span>
<span><span class="co">#&gt; $warnings[[1]]</span></span>
<span><span class="co">#&gt; $warnings[[1]][[1]]</span></span>
<span><span class="co">#&gt; [1] "2"</span></span>
<span><span class="co">#&gt; </span></span>
<span><span class="co">#&gt; $warnings[[1]][[2]]</span></span>
<span><span class="co">#&gt; [1] "age"</span></span>
<span><span class="co">#&gt; </span></span>
<span><span class="co">#&gt; $warnings[[1]][[3]]</span></span>
<span><span class="co">#&gt; [1] "On row 2 the attribute age does not contain the proper value type num."</span></span>
<span><span class="co">#&gt; </span></span>
<span><span class="co">#&gt; $warnings[[1]][[4]]</span></span>
<span><span class="co">#&gt; [1] ""</span></span>
<span><span class="co">#&gt; </span></span>
<span><span class="co">#&gt; </span></span>
<span><span class="co">#&gt; $warnings[[2]]</span></span>
<span><span class="co">#&gt; $warnings[[2]][[1]]</span></span>
<span><span class="co">#&gt; [1] "3"</span></span>
<span><span class="co">#&gt; </span></span>
<span><span class="co">#&gt; $warnings[[2]][[2]]</span></span>
<span><span class="co">#&gt; [1] "age"</span></span>
<span><span class="co">#&gt; </span></span>
<span><span class="co">#&gt; $warnings[[2]][[3]]</span></span>
<span><span class="co">#&gt; [1] "On row 3 the attribute age does not contain the proper value type num."</span></span>
<span><span class="co">#&gt; </span></span>
<span><span class="co">#&gt; $warnings[[2]][[4]]</span></span>
<span><span class="co">#&gt; [1] ""</span></span>
<span><span class="co">#&gt; </span></span>
<span><span class="co">#&gt; </span></span>
<span><span class="co">#&gt; $warnings[[3]]</span></span>
<span><span class="co">#&gt; $warnings[[3]][[1]]</span></span>
<span><span class="co">#&gt; $warnings[[3]][[1]][[1]]</span></span>
<span><span class="co">#&gt; [1] "2"</span></span>
<span><span class="co">#&gt; </span></span>
<span><span class="co">#&gt; $warnings[[3]][[1]][[2]]</span></span>
<span><span class="co">#&gt; [1] "3"</span></span>
<span><span class="co">#&gt; </span></span>
<span><span class="co">#&gt; </span></span>
<span><span class="co">#&gt; $warnings[[3]][[2]]</span></span>
<span><span class="co">#&gt; [1] "readPair"</span></span>
<span><span class="co">#&gt; </span></span>
<span><span class="co">#&gt; $warnings[[3]][[3]]</span></span>
<span><span class="co">#&gt; [1] "readPair values in rows ['2', '3'] are out of the specified range."</span></span>
<span><span class="co">#&gt; </span></span>
<span><span class="co">#&gt; $warnings[[3]][[4]]</span></span>
<span><span class="co">#&gt; $warnings[[3]][[4]][[1]]</span></span>
<span><span class="co">#&gt; [1] ""</span></span>
<span><span class="co">#&gt; </span></span>
<span><span class="co">#&gt; </span></span>
<span><span class="co">#&gt; </span></span>
<span><span class="co">#&gt; $warnings[[4]]</span></span>
<span><span class="co">#&gt; $warnings[[4]][[1]]</span></span>
<span><span class="co">#&gt; [1] "2"</span></span>
<span><span class="co">#&gt; </span></span>
<span><span class="co">#&gt; $warnings[[4]][[2]]</span></span>
<span><span class="co">#&gt; [1] "readLength"</span></span>
<span><span class="co">#&gt; </span></span>
<span><span class="co">#&gt; $warnings[[4]][[3]]</span></span>
<span><span class="co">#&gt; [1] "On row 2 the attribute readLength does not contain the proper value type int."</span></span>
<span><span class="co">#&gt; </span></span>
<span><span class="co">#&gt; $warnings[[4]][[4]]</span></span>
<span><span class="co">#&gt; [1] ""</span></span>
<span><span class="co">#&gt; </span></span>
<span><span class="co">#&gt; </span></span>
<span><span class="co">#&gt; $warnings[[5]]</span></span>
<span><span class="co">#&gt; $warnings[[5]][[1]]</span></span>
<span><span class="co">#&gt; [1] "3"</span></span>
<span><span class="co">#&gt; </span></span>
<span><span class="co">#&gt; $warnings[[5]][[2]]</span></span>
<span><span class="co">#&gt; [1] "readLength"</span></span>
<span><span class="co">#&gt; </span></span>
<span><span class="co">#&gt; $warnings[[5]][[3]]</span></span>
<span><span class="co">#&gt; [1] "On row 3 the attribute readLength does not contain the proper value type int."</span></span>
<span><span class="co">#&gt; </span></span>
<span><span class="co">#&gt; $warnings[[5]][[4]]</span></span>
<span><span class="co">#&gt; [1] ""</span></span>
<span><span class="co">#&gt; </span></span>
<span><span class="co">#&gt; </span></span>
<span><span class="co">#&gt; $warnings[[6]]</span></span>
<span><span class="co">#&gt; $warnings[[6]][[1]]</span></span>
<span><span class="co">#&gt; [1] "2"</span></span>
<span><span class="co">#&gt; </span></span>
<span><span class="co">#&gt; $warnings[[6]][[2]]</span></span>
<span><span class="co">#&gt; [1] "readDepth"</span></span>
<span><span class="co">#&gt; </span></span>
<span><span class="co">#&gt; $warnings[[6]][[3]]</span></span>
<span><span class="co">#&gt; [1] "On row 2 the attribute readDepth does not contain the proper value type int."</span></span>
<span><span class="co">#&gt; </span></span>
<span><span class="co">#&gt; $warnings[[6]][[4]]</span></span>
<span><span class="co">#&gt; [1] ""</span></span>
<span><span class="co">#&gt; </span></span>
<span><span class="co">#&gt; </span></span>
<span><span class="co">#&gt; $warnings[[7]]</span></span>
<span><span class="co">#&gt; $warnings[[7]][[1]]</span></span>
<span><span class="co">#&gt; [1] "3"</span></span>
<span><span class="co">#&gt; </span></span>
<span><span class="co">#&gt; $warnings[[7]][[2]]</span></span>
<span><span class="co">#&gt; [1] "readDepth"</span></span>
<span><span class="co">#&gt; </span></span>
<span><span class="co">#&gt; $warnings[[7]][[3]]</span></span>
<span><span class="co">#&gt; [1] "On row 3 the attribute readDepth does not contain the proper value type int."</span></span>
<span><span class="co">#&gt; </span></span>
<span><span class="co">#&gt; $warnings[[7]][[4]]</span></span>
<span><span class="co">#&gt; [1] ""</span></span>
<span><span class="co">#&gt; </span></span>
<span><span class="co">#&gt; </span></span>
<span><span class="co">#&gt; $warnings[[8]]</span></span>
<span><span class="co">#&gt; $warnings[[8]][[1]]</span></span>
<span><span class="co">#&gt; [1] "2"</span></span>
<span><span class="co">#&gt; </span></span>
<span><span class="co">#&gt; $warnings[[8]][[2]]</span></span>
<span><span class="co">#&gt; [1] "experimentalTimepoint"</span></span>
<span><span class="co">#&gt; </span></span>
<span><span class="co">#&gt; $warnings[[8]][[3]]</span></span>
<span><span class="co">#&gt; [1] "On row 2 the attribute experimentalTimepoint does not contain the proper value type num."</span></span>
<span><span class="co">#&gt; </span></span>
<span><span class="co">#&gt; $warnings[[8]][[4]]</span></span>
<span><span class="co">#&gt; [1] ""</span></span>
<span><span class="co">#&gt; </span></span>
<span><span class="co">#&gt; </span></span>
<span><span class="co">#&gt; $warnings[[9]]</span></span>
<span><span class="co">#&gt; $warnings[[9]][[1]]</span></span>
<span><span class="co">#&gt; [1] "3"</span></span>
<span><span class="co">#&gt; </span></span>
<span><span class="co">#&gt; $warnings[[9]][[2]]</span></span>
<span><span class="co">#&gt; [1] "experimentalTimepoint"</span></span>
<span><span class="co">#&gt; </span></span>
<span><span class="co">#&gt; $warnings[[9]][[3]]</span></span>
<span><span class="co">#&gt; [1] "On row 3 the attribute experimentalTimepoint does not contain the proper value type num."</span></span>
<span><span class="co">#&gt; </span></span>
<span><span class="co">#&gt; $warnings[[9]][[4]]</span></span>
<span><span class="co">#&gt; [1] ""</span></span></code></pre></div>
<ol start="3" style="list-style-type: decimal">
<li><p>Make corrections in the <code>.csv</code> according to validation
laundry list.</p></li>
<li><p>Submit corrected manifest via DCA.</p></li>
</ol>
</div>
<div class="section level2">
<h2 id="alternative-with-dataset-entity">Alternative with dataset entity<a class="anchor" aria-label="anchor" href="#alternative-with-dataset-entity"></a>
</h2>
<p>For working with dataset entities or doing extra checks, this
slightly different workflow can be applied instead. Dataset entities may
be more complicated because they can combine files in different times
and places (batches).</p>
<ol style="list-style-type: decimal">
<li>Generate a manifest for Synapse dataset entity. We need to use
<code>remanifest</code>; to understand differences, read the docs by
running <code><a href="../reference/remanifest.html">?remanifest</a></code>.</li>
</ol>
<div class="sourceCode" id="cb4"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">datasets</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/list_project_datasets.html">list_project_datasets</a></span><span class="op">(</span>project_id <span class="op">=</span> <span class="st">"syn4939902"</span>, type <span class="op">=</span> <span class="st">"dataset"</span><span class="op">)</span></span>
<span><span class="va">new_datasets</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/funprog.html" class="external-link">Filter</a></span><span class="op">(</span><span class="kw">function</span><span class="op">(</span><span class="va">d</span><span class="op">)</span> <span class="fu"><a href="https://rdrr.io/r/base/as.Date.html" class="external-link">as.Date</a></span><span class="op">(</span><span class="va">d</span><span class="op">$</span><span class="va">createdOn</span><span class="op">)</span> <span class="op">&gt;</span> <span class="fu"><a href="https://rdrr.io/r/base/as.Date.html" class="external-link">as.Date</a></span><span class="op">(</span><span class="st">"2023-12-01"</span><span class="op">)</span>, <span class="va">datasets</span><span class="op">)</span> <span class="co"># or filter by name</span></span></code></pre></div>
<p>We will use the first item as the demo.</p>
<div class="sourceCode" id="cb5"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">test</span> <span class="op">&lt;-</span> <span class="va">new_datasets</span><span class="op">[[</span><span class="fl">1</span><span class="op">]</span><span class="op">]</span><span class="op">$</span><span class="va">id</span></span>
<span><span class="fu"><a href="../reference/remanifest.html">remanifest</a></span><span class="op">(</span><span class="va">test</span>, file <span class="op">=</span> <span class="st">"manifest_rd1.csv"</span><span class="op">)</span></span>
<span><span class="co">#&gt; ✔️ Saved manifest as manifest_rd1.csv</span></span></code></pre></div>
<p>For reproducibility this original manifest “manifest_rd1.csv” is in
the vignettes folder.</p>
<ol start="2" style="list-style-type: decimal">
<li>Run precheck.</li>
</ol>
<div class="sourceCode" id="cb6"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="../reference/precheck_manifest.html">precheck_manifest</a></span><span class="op">(</span><span class="st">"manifest_rd1.csv"</span><span class="op">)</span></span>
<span><span class="co">#&gt; ❌ Multiple components detected in a single manifest: 'RNASeqTemplate', 'GenomicsAssayTemplate', ''. This can happen when files were annotated at different eras.</span></span>
<span><span class="co">#&gt; Suggestions: 1) Split up the manifest because schematic can only validate one type at a type. 2) Harmonize the components if this is sensible.</span></span>
<span><span class="co">#&gt; For example, RNASeqTemplate is an alias for GenomicsAssayTemplate</span></span>
<span><span class="co">#&gt; ❌ Blank value '' for Component detected. This can happen because files were annotated before 2022, when Component was introduced for most DCCs.</span></span>
<span><span class="co">#&gt; ❌ The pattern of these attribute names suggest duplicates: '...1', '...44', '...46'. This may happen when metadata is supplemented programmatically with a data-type mismatch</span></span>
<span><span class="co">#&gt; ⚠️ An attribute `Uuid` is present and should preferably be removed. See issue # .</span></span>
<span><span class="co">#&gt; ⚠️ An attribute `eTag` is present and preferably be removed.</span></span>
<span><span class="co">#&gt; ℹ️ Custom attributes (not documented in data model) were found: 'entityId', '...1', 'Uuid', 'DNA_ID', 'RNA_ID', 'tissue', 'bodyPart', 'parentSpecimenId', 'eTag', '...44', '...46', 'accessTeam', 'accessType', 'sciDataRelease', 'specimenIdSource', 'timePointUnit', 'transplantationDonorTissue', 'transplantationDonorSpecies'. In general, custom attributes added by the researcher to help with data management are fine.</span></span>
<span><span class="co">#&gt; Just check that they are not PHI or added by mistake. If they are deemed generally useful or important enough, they can also be documented officially in the data model for others to reference.</span></span></code></pre></div>
<ol start="3" style="list-style-type: decimal">
<li><p>Make a copy of “manifest_rd1.csv”,
e.g. “manifest_rd1_corrected.csv” and use precheck notes to help make
corrections.</p></li>
<li><p>Validate “manifest_rd1_corrected.csv” using schematic service
(same as above, not run since general validation output has already been
shown).</p></li>
<li><p>Make <em>more</em> corrections as needed.</p></li>
<li><p>Finally submit corrected manifest.</p></li>
</ol>
</div>
  </main><aside class="col-md-3"><nav id="toc"><h2>On this page</h2>
    </nav></aside>
</div>



    <footer><div class="pkgdown-footer-left">
  <p></p>
<p>Developed by Robert Allaway, Anh Nguyet Vu.</p>
</div>

<div class="pkgdown-footer-right">
  <p></p>
<p>Site built with <a href="https://pkgdown.r-lib.org/" class="external-link">pkgdown</a> 2.0.7.</p>
</div>

    </footer>
</div>

  

  

  </body>
</html>
